tests/handlers/test_directory.py

```diff
@@ -432,6 +432,48 @@ def test_allowed(self):
         self.assertEquals(200, channel.code, channel.result)
 
 
+class TestCreatePublishedRoomACL(unittest.HomeserverTestCase):
+    user_id = "@test:test"
+    data_template = (
+        '{"room_alias_name": "%%23unofficial_test%%3Atest", "visibility": "%s"}'
+    )
+
+    servlets = [directory.register_servlets, room.register_servlets]
+
+    def prepare(self, reactor, clock, hs):
+        # This time we add custom room list publication rules
+        config = {}
+        config["alias_creation_rules"] = []
+        config["room_list_publication_rules"] = [
+            {"user_id": "*", "alias": "*", "action": "deny"}
+        ]
+
+        rd_config = RoomDirectoryConfig()
+        rd_config.read_config(config)
+
+        self.hs.config.roomdirectory.is_publishing_room_allowed = (
+            rd_config.is_publishing_room_allowed
+        )
+
+        return hs
+
+    def test_denied(self):
+        channel = self.make_request(
+            "POST", "createRoom", (self.data_template % ("public",)).encode("ascii")
+        )
+        self.assertEquals(403, channel.code, channel.result)
```

Comment:
I don't think this test would fail before your changes. I think we'd need to add logic which checks that we haven't written to the DB. Let me see if I can find an example...

Reply:
Here's an example where we select columns from a table with a where clause, and assert that we selected no rows:

https://github.com/matrix-org/synapse/blob/dc75fb7f0552e6a9903a7c173672c96610219ec0/tests/storage/databases/main/test_room.py#L57-L65

We'd have to work out what table is being written to and so on.

Alternatively, we could follow your description on #10929 and try to create a room with the same alias, but as a user with permission. That user should succeed in doing so.

> Try to create the same room again, but with a different user that has permission to publish rooms
> Attempts to use the same alias for the new room will fail, since the alias is still in use for the room that failed to be created

That would involve 

- `self.register_user` to make a new user
- `self.login` to generate an access token for that user
- passing `access_token=...` to `self.make_request` calls to act as that user.

Here's an example:

https://github.com/matrix-org/synapse/blob/e584534403b55ad3f250f92592e30b15b01f0201/tests/rest/admin/test_room.py#L1035-L1045



Sorry, that's quite an info dump. How does that feel---does it make sense/feel doable?

Reply:
I added a test to create a published room as a denied, then allowed, user.

I managed to do it just by setting a custom `user_id` in calls to `self.helper.create_room_as`, while the test context is managed by an admin user. It seems to work fine, but please let me know if what I did is problematic.